Mage.Sets/src/mage/cards/e/EvolvingAdaptive.java

```diff
@@ -0,0 +1,95 @@
+package mage.cards.e;
+
+import mage.MageInt;
+import mage.abilities.common.*;
+import mage.abilities.dynamicvalue.DynamicValue;
+import mage.abilities.dynamicvalue.common.CountersSourceCount;
+import mage.abilities.effects.Effect;
+import mage.abilities.effects.common.continuous.BoostSourceEffect;
+import mage.abilities.effects.common.counter.AddCountersSourceEffect;
+import mage.cards.CardImpl;
+import mage.cards.CardSetInfo;
+import mage.constants.CardType;
+import mage.constants.Duration;
+import mage.constants.SubType;
+import mage.counters.CounterType;
+import mage.filter.FilterPermanent;
+import mage.filter.common.FilterCreaturePermanent;
+import mage.filter.predicate.mageobject.AnotherPredicate;
+import mage.game.Game;
+import mage.game.events.GameEvent;
+import mage.game.permanent.Permanent;
+
+import java.util.UUID;
+
+public class EvolvingAdaptive extends CardImpl {
+
+    private static final FilterPermanent filter = new FilterCreaturePermanent("another creature");
+
+    static {
+        filter.add(AnotherPredicate.instance);
+    }
+
+    private static final DynamicValue oilCounters = new CountersSourceCount(CounterType.OIL);
+
+    public EvolvingAdaptive(UUID ownerId, CardSetInfo setInfo) {
+        super(ownerId, setInfo, new CardType[]{CardType.CREATURE}, "{G}");
+        this.addSubType(SubType.PHYREXIAN);
+        this.addSubType(SubType.WARRIOR);
+        this.power = new MageInt(0);
+        this.toughness = new MageInt(0);
+
+        //Evolving Adaptive enters the battlefield with an oil counter on it.
+        this.addAbility(new EntersBattlefieldAbility(
+                new AddCountersSourceEffect(CounterType.OIL.createInstance()),
+                "with an oil counter on it"
+        ));
+
+        //Evolving Adaptive gets +1/+1 for each oil counter on it.
+        this.addAbility(new SimpleStaticAbility(new BoostSourceEffect(oilCounters, oilCounters, Duration.WhileOnBattlefield)
+                .setText("{this} gets +1/+1 for each oil counter on it")));
+
+        //Whenever another creature enters the battlefield under your control, if that creature has greater power or
+        //toughness than Evolving Adaptive, put an oil counter on Evolving Adaptive.
+        this.addAbility(new EvolvingAdaptiveTriggeredAbility(new AddCountersSourceEffect(CounterType.OIL.createInstance())
+                .setText("if that creature has greater power or toughness than Evolving Adaptive, put an oil counter " +
+                        "on Evolving Adaptive.")
+                , filter, this.objectId));
+    }
+
+    private EvolvingAdaptive(final EvolvingAdaptive card) {
+        super(card);
+    }
+
+    @Override
+    public EvolvingAdaptive copy() {
+        return new EvolvingAdaptive(this);
+    }
+}
+
+class EvolvingAdaptiveTriggeredAbility extends EntersBattlefieldControlledTriggeredAbility {
+
+    UUID permanentID;
+
+    public EvolvingAdaptiveTriggeredAbility(Effect effect, FilterPermanent filter, UUID permanentID) {
```

Comment:
You do not need to pass all of these parameters to the constructor. The effect can be specified within this ability (as opposed to being passed in from `EvolvingAdaptive`), the filter can be specified within this ability (again, as opposed to being passed in from `EvolvingAdaptive`), and you only use the `UUID permanentID` to get the source permanent on line 82 when you write:
```java
Permanent permanent = game.getPermanent(permanentID);
```
However, this is also unnecessary as you can achieve the same thing without passing in an ID by doing:
```java
Permanent permanent = getSourcePermanentIfItStillExists(game);
```
See [Pelt Collector](https://github.com/magefree/mage/blob/754a5b38ebba104b83494629c1f3881020168b1e/Mage.Sets/src/mage/cards/p/PeltCollector.java) as an example of what I mean about not needing to pass parameters into a custom ability like this one

Reply:
hopefully done ;)